### PR TITLE
Inline JIT C++ Extensions

### LINKED ISSUE #7012: [feature request] [pytorch] Support inline cpp/cuda JIT extensions
If I understand correctly, currently the JIT variant requires the C++/CUDA files to live on disk. It would be cool to support compiling C++/CUDA code from a string variable from a python variable, like in nvrtc/cupy: https://github.com/vadimkantorov/caffemodel2pytorch (ROI pooling example).

This would reduce boiler-plate files and improve code clarity for small CUDA functions.

### PR DESCRIPTION
**Pull request by goldsborough**
Adds ability to JIT compile C++ extensions from strings

```py
>>> from torch.utils.cpp_extension import load_inline
>>> source = '''
    at::Tensor sin_add(at::Tensor x, at::Tensor y) {
      return x.sin() + y.sin();
    }
'''
>>> module = load_inline(name='inline_extension', cpp_sources=source, functions='sin_add')
```

Fixes #7012 

@apaszke @fmassa

### CODE REVIEW

File: test/test_cpp_extensions.py [hunk 1]
```diff
@@ -106,6 +106,83 @@ def test_optional(self):
...
+        #include <torch/torch.h>
+        at::Tensor sin_add(at::Tensor x, at::Tensor y) {
+          return x.sin() + y.sin();
+        }
+        '''
+
+        cpp_source2 = '''
+        #include <torch/torch.h>
+        at::Tensor sin_add(at::Tensor x, at::Tensor y);
+        PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
```
Comment (fmassa) [inline test/test_cpp_extensions.py]: A random idea: it could be possible to auto generate the `PYBIND11_MODULE` by letting the user specify a list of tuples (or dict) in python, that define the necessary arguments for pybind.

For example, if we have a list of dicts to be passed as a `signatures` keyword to `load_inline`, the user would be able to just call with  `signatures=[{'name': 'sin_add', 'doc': 'sin(a) + sin(b)'}]` and let `cname` be inferred from `name`.

What do you think?

File: torch/utils/cpp_extension.py [hunk 2]
```diff
@@ -461,20 +461,95 @@ def load(name,
...
+    identical to :func:`load`. Strings passed in ``cpp_sources`` (a string or
+    list of strings) are stored with a `.cpp` extension, and the string or list
+    of strings passed in ``cuda_sources`` are stored with a ``.cu`` extension.
 
-    if build_directory is None:
-        build_directory = _get_build_directory(name, verbose)
+    Example:
+        >>> from torch.utils.cpp_extension import load_inline
+        >>> source = \'\'\'
+        #include <torch/torch.h>
```
Comment (fmassa) [inline torch/utils/cpp_extension.py]: Maybe we could append this by default in the string passed to the user?

File: test/test_cpp_extensions.py [hunk 3]
```diff
@@ -106,6 +106,83 @@ def test_optional(self):
...
+          return output;
+        }
+        PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
+          m.def("cos_add", &cos_add, "cos(x) + cos(y)");
+        }
+        '''
+
+        cuda_source = '''
+        #include <cuda.h>
+        #include <cuda_runtime.h>
```
Comment (fmassa) [inline test/test_cpp_extensions.py]: Do we actually need those cuda includes? I think ATen already includes them, right?

**Review by fmassa — APPROVED** [review]
I think this looks very good!

I added a new minor comments. I was thinking, given that the user can specify the strings on python and call the generation from python, we could maybe relax a box the number of things that the user need to write.

What do you think?

File: test/test_cpp_extensions.py (hunk 3 above)
Comment (goldsborough) [reply test/test_cpp_extensions.py]: No we never include `cuda.h` in headers, only in implementation files

File: torch/utils/cpp_extension.py (hunk 2 above)
Comment (goldsborough) [reply torch/utils/cpp_extension.py]: Yes I think that's a good idea

File: test/test_cpp_extensions.py (hunk 1 above)
Comment (goldsborough) [reply test/test_cpp_extensions.py]: Yes definitely -- maybe even just a list of function names should work. I imagine often folks won't care about the docstring. So `signatures` should take a list or dictionary

Comment (goldsborough) [reply test/test_cpp_extensions.py]: It's a bit tricky because we need to pick which `.cpp` source to put the `PYBIND11_MODULE` into. We'll just use the first one

Comment (fmassa) [reply test/test_cpp_extensions.py]: You could just write everything the user pass as a `.hpp` file, and import them all in a `main.cpp` that contains the `PYBIND11_MODULE`, right?

Comment (goldsborough) [reply test/test_cpp_extensions.py]: True. Now that I think of it, I'm not sure it actually makes sense that `cpp_sources` and `cuda_sources` be a list of strings. I think it would be enough that there is one `cpp_source` and one` cuda_source`, what do you think?

File: test/test_cpp_extensions.py (hunk 3 above)
Comment (fmassa) [reply test/test_cpp_extensions.py]: Oh, I see. I didn't need to add those because I included in my files `<THC/THC.h>`, which [includes `cuda.h` and `cuda_runtime.h`](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCGeneral.h.in#L12-L13).

File: test/test_cpp_extensions.py (hunk 1 above)
Comment (fmassa) [reply test/test_cpp_extensions.py]: I think any of them is fine.
I could imagine people writing individual functions in strings, in which case having the list of sources makes sense.

**Comment by goldsborough** [comment]
@fmassa check out the tests to see what it looks like now. Instead of putting each source into a header and including it in `main.cpp`, I just concatenate the sources into a `main.cpp` and a `cuda.cu`. This is equivalent since `#include <...>` is basically concatenating sources :)

Let me know what you think and I'll update the docs

File: test/test_cpp_extensions.py [hunk 4]
```diff
@@ -106,6 +106,111 @@ def test_optional(self):
...
+        at::Tensor sin_add(at::Tensor x, at::Tensor y) {
+          return x.sin() + y.sin();
+        }
+        '''
+
+        cpp_source2 = '''
+        #include <torch/torch.h>
+        at::Tensor sin_add(at::Tensor x, at::Tensor y);
+        PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
+          m.def("sin_add", &sin_add, "sin(x) + sin(y)");
```
Comment (ezyang) [inline test/test_cpp_extensions.py]: This is a bit late to ask this question (since the code for this has already been merged) but I'll ask it anyway: are we sure we want pybind11 to be part of the public interface of C++ extensions? If it is, then we are (1) always on the hook to ship pybind11 as part of PyTorch (we cannot ever decide "oh, let's use another binding library", because that will break userspace), and (2) every upgrade of pybind11 is a potentially user-visible BC-breaking change, because any BC-breakage in pybind11 is BC breakage for extension code. (Similarly, it looks like we are installing pybind11 headers as part of PyTorch, which is not very good OSS citizenship, because a downstream project might want to be using their own version of pybind11, and the headers will clobber each other.)

If we don't have concrete use-cases for letting people directly write pybind11 in the extension code, then it seems safer to (1) move our installed pybind11 headers under a private prefix (so they don't clash with a user) and (2) don't expose the functionality to directly write bindings. It's not a big loss; after all, the user can use whatever they want to get bindings, if they really want to.

File: torch/utils/cpp_extension.py [hunk 5]
```diff
@@ -461,20 +461,119 @@ def load(name,
...
+        >>> source = \'\'\'
+        at::Tensor sin_add(at::Tensor x, at::Tensor y) {
+          return x.sin() + y.sin();
+        }
+        \'\'\'
+        >>> module = load_inline(name='inline_extension',
+                                 cpp_sources=[source],
+                                 functions=['sin_add'])
+    '''
+    build_directory = build_directory or _get_build_directory(name, verbose)
```
Comment (ezyang) [inline torch/utils/cpp_extension.py]: How discoverable is the implicitly assigned `_get_build_directory`? Users find it very helpful to be able to find where a source file is being built if they're debugging some compile problem.

Comment (ezyang) [reply torch/utils/cpp_extension.py]: An especially poignant case is if the automatically generated pybind11 code isn't quite working.

**Review by ezyang — APPROVED** [review]

File: test/test_cpp_extensions.py (hunk 4 above)
Comment (goldsborough) [reply test/test_cpp_extensions.py]: Zach and I consciously decided to expose pybind11 instead of some kind of custom `TORCH_EXTENSION` so that users could answer all their questions about bindings by looking at pybind11 docs instead of having to ask us. This makes it clear that we are not doing anything magic here, it's just pybind11. 

Your concerns are very valid but I feel like they would affect 1 out of 1,000,000 users and make the experience less pleasant for the other 999,999? For example, the extension tests, which we point to as examples of using the extension API, also have examples of binding in structs. This is even more advanced usage of pybind11, which we would have to wrap somehow, or disallow.

File: torch/utils/cpp_extension.py (hunk 5 above)
Comment (goldsborough) [reply torch/utils/cpp_extension.py]: If you set `verbose=True` it gets printed

**Review by apaszke — APPROVED** [review]
Looks nice! It would be great if we could somehow spare specifying the function signature in C++ sources just to get this thing to compile, but it doesn't seem too easy.

**Review by fmassa — APPROVED** [review]
That looks great, thanks @goldsborough !
Let's just update the docs for this function and get this merged

**Comment by goldsborough** [comment]
@ezyang I still need to update docs 😄
